### PR TITLE
Generate transient dependencies list

### LINKED ISSUE #1824: Bump transitive dependencies in "netfx" to higher version
As explained in this [comment](https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/issues/1646#issuecomment-1349083888) with the implementation of #1646 there are good reasons to bump up the transitive dependencies of `OpenTelemetry.AutoInstrumentation` to their latest version. Unfortunately, there are no straightforward automated ways of doing that, this issue is to automate this transitive version bump as part of the project build.

### PR DESCRIPTION
**Pull request by RassK**
## Why

Fixes #1824

## What

1. Generates transient dependencies list to `OpenTelemetry.AutoInstrumentation`
2. Dependabot will start bumping the versions
3. A test verifies that a PR is not bringing new undefined transient dependencies

## Tests

Verifies transient deps.

## Checklist

- [x] New features are covered by tests.

### CODE REVIEW

File: build/Build.Steps.Windows.cs [hunk 1]
```diff
@@ -118,6 +120,50 @@ partial class Build
             );
         });
 
+    Target GenerateNetFxTransientDependencies => _ => _
+        .Unlisted()
```
Comment (pellared) [inline build/Build.Steps.Windows.cs:124]: This target is unlisted and not used anywhere. Shouldn't it be invoked in GH Actions? Otherwise we will have missing entries. Personally I think we could add it to the `Workflow`. We have a `git diff` in `ci.yml - build` GH job that would make sure that everything is up to date.

File: build/Build.Steps.Windows.cs [hunk 2]
```diff
@@ -13,8 +15,8 @@ partial class Build
 {
     Target CompileNativeSrcWindows => _ => _
         .Unlisted()
+        .DependsOn(GenerateNetFxAssemblyRedirectionSource)
```
Comment (pellared) [inline build/Build.Steps.Windows.cs:18]: Why this change? I do not remember the exact details right now, but there were some reasons that we are not using the `DependsOn` in "step targets". Reference: https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/issues/889

**Comment by pjanotti** [comment]
@rajkumar-rangaraj perhaps we could use a similar approach to pull the transitive dependencies independent of the TFM that we use for the `OpenTelemetry.AutoInstrumentation`. This will ensure the presence of the needed assemblies while keeping targeting the TFM provided by the SDK. Just an idea that keeps things more under our own control.

File: build/Build.Steps.Windows.cs (hunk 1 above)
Comment (pjanotti) [reply build/Build.Steps.Windows.cs:124]: I agree with this: unless it is relatively slow I would say that it should be executed before the `CompileManagedSrc` when the TFM is `net462`. If it is really slow it could be part of the `Workflow`.

**Comment by rajkumar-rangaraj** [comment]
> @rajkumar-rangaraj perhaps we could use a similar approach to pull the transitive dependencies independent of the TFM that we use for the `OpenTelemetry.AutoInstrumentation`. This will ensure the presence of the needed assemblies while keeping targeting the TFM provided by the SDK. Just an idea that keeps things more under our own control.

@pjanotti I did not understand the context, could you give little background.

File: build/Build.Steps.Windows.cs (hunk 1 above)
Comment (RassK) [reply build/Build.Steps.Windows.cs:124]: I currently don't see any need to auto execute it. It should be executed only within a PR that is going to break dependencies.
Secondly it will generate uncommitted changes then, but all of these changes must be committed for dependendabot to work.

File: build/Build.Steps.Windows.cs (hunk 2 above)
Comment (RassK) [reply build/Build.Steps.Windows.cs:18]: This will decouple `GenerateNetFxAssemblyRedirectionSource` from `CompileNativeSrcWindows` imo?
`GenerateNetFxAssemblyRedirectionSource` is only connected to `CompileNativeSrcWindows`...

File: build/Build.Steps.Windows.cs (hunk 1 above)
Comment (pellared) [reply build/Build.Steps.Windows.cs:124]: > It should be executed only within a PR that is going to break dependencies.

How we will know it if it is not automated?

> Secondly it will generate uncommitted changes then, but all of these changes must be committed for dependendabot to work.

Then we will just push commits to the dependabot PRs. Thanks to it we will make sure that `main` is always "clean".

File: build/Build.Steps.Windows.cs (hunk 2 above)
Comment (pellared) [reply build/Build.Steps.Windows.cs:18]: `BuildTracer` is the "workflow" target which has the `DependsOn`. I propose to follow the current pattern otherwise we will have a mess.

File: build/Build.Steps.Windows.cs (hunk 1 above)
Comment (pjanotti) [reply build/Build.Steps.Windows.cs:124]: The goal is to have the list of assemblies automatically update whenever needed. It seems to me that the simplest way to achieve that seems to have this step as part of `BuildTracer` target.

**Comment by pjanotti** [comment]
@rajkumar-rangaraj I mean in the context that currently we build the tracer targeting `net6.0`, however, since some transitive dependencies don't have packages for `net6.0` we don't get the assemblies of such transitive dependencies under the `net`. I did some quick experiments changing the TFM to `netstandard2.0` and this improved the situation. However, as you mentioned, the SDK is removing the `netstandard2.0` target, which makes sense, since they are targeted to applications/exes, that will pull all the needed dependencies at build time.

For automatic instrumentation, it __seems__ that we need to pull all these references for cases when the AdditionalDeps or SharedStore can't be used. I still want to investigate this further and confirm it. Anyway, if we do need to pull these assemblies using the tool being added in this PR could pull the needed assemblies. Makes sense?

File: build/Build.Steps.Windows.cs (hunk 1 above)
Comment (RassK) [reply build/Build.Steps.Windows.cs:124]: > How we will know it if it is not automated?
That's exactly why the test is there. It enforces you to run `GenerateNetFxTransientDependencies` just in the right time, so external dependencies are clearly visible and connected to the PR that caused these. 
Running it fully automated causes dependencies to be added without any clear source trail.

File: build/Build.Steps.Windows.cs (hunk 2 above)
Comment (RassK) [reply build/Build.Steps.Windows.cs:18]: Which is exactly my point. I'm running these sub targets quite a lot. If you do clean and run `CompileNativeSrc` it won't compile due missing source file. Which means `CompileNativeSrc` does not do what it promises.

File: build/Build.Steps.Windows.cs (hunk 1 above)
Comment (pjanotti) [reply build/Build.Steps.Windows.cs:124]: A test covering this is good too but as of right now if the csproj gets updated due to some other change CI will flag it so no accidental changes are going to go through one way or another. If you have it automated as part of the build the change will be visible locally even before the file is submitted so the author can review it.

File: build/Build.Steps.Windows.cs (hunk 2 above)
Comment (pjanotti) [reply build/Build.Steps.Windows.cs:18]: In order to run `CompileNativeSrc` and get the correct bits under `tracer-home` one has to `BuildTracer` before (likely after a clean). My workflow is `BuildTracer` once and then build the components as needed, in this case, it will be nice to avoid the new target when doing `CompileNativeSrc`.

Comment (RassK) [reply build/Build.Steps.Windows.cs:18]: I'll revert this part, not tightly coupled to this PR.

But in general I still don't agree. `CompileNativeSrcWindows` should do exactly what it promises, otherwise it's something else. If you need to optimize `CompileNativeSrcWindows` you can always do `nuke CompileNativeSrcWindows -skip GenerateNetFxAssemblyRedirectionSource` which then skips the parts and does personalized run.

Comment (pellared) [reply build/Build.Steps.Windows.cs:18]: `nuke CompileNativeSrcWindows -skip` would even work

Comment (pellared) [reply build/Build.Steps.Windows.cs]: I added https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/issues/889#issuecomment-1406277116

File: build/Build.Steps.Windows.cs (hunk 1 above)
Comment (RassK) [reply build/Build.Steps.Windows.cs:124]: That actually makes sense, I originally thought that the target is somewhat expensive to be in the main pipeline. 
On my desktop it's just 2 seconds (looking at nuke's feedback), will see after a CI run as it's the weakest machine.

Comment (RassK) [reply build/Build.Steps.Windows.cs:124]: well `GenerateNetFxTransientDependencies Succeeded  0:13` not too terrible actually.

**Review by pellared — APPROVED** [review]

File: build/Build.Steps.Windows.cs (hunk 1 above)
Comment (pjanotti) [reply build/Build.Steps.Windows.cs:124]: Excellent! Much better than what I expected too.

**Review by pjanotti — APPROVED** [review]
